jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -728,55 +736,55 @@ protected void writeHttpContent(Request request, Response response, Callback cal
-//        headers.remove(EnumSet.of(
-//            HttpHeader.LAST_MODIFIED,
-//            HttpHeader.CONTENT_LENGTH,
-//            HttpHeader.CONTENT_TYPE,
-//            HttpHeader.CONTENT_ENCODING,
-//            HttpHeader.ETAG,
-//            HttpHeader.ACCEPT_RANGES,
-//            HttpHeader.CACHE_CONTROL
-//            ));
-//        HttpField lm = content.getLastModified();
-//        if (lm != null)
-//            headers.add(lm);
-//        etc.
+        HttpFields.Mutable headers = response.getHeaders();
 
+        // Existing etags have priority over content etags (often set by compression handler)
+        if (_etags && !headers.contains(HttpHeader.ETAG))
+        {
+            HttpField et = content.getETag();
+            if (et != null)
+                headers.add(et);
+        }
+
+        // Existing content encoding is kept only if not set for content.
+        HttpField ce = content.getContentEncoding();
+        if (ce != null)
+            headers.put(ce);
+
+        // Remove content headers and re-add if we have them
+        headers.remove(CONTENT_HEADERS);
         HttpField lm = content.getLastModified();
         if (lm != null)
-            response.getHeaders().put(lm);
-
+            headers.add(lm);
         if (contentLength == USE_KNOWN_CONTENT_LENGTH)
-        {
-            response.getHeaders().put(content.getContentLength());
-        }
+            headers.add(content.getContentLength());
         else if (contentLength > NO_CONTENT_LENGTH)
-        {
-            response.getHeaders().put(HttpHeader.CONTENT_LENGTH, contentLength);
-        }
-
+            headers.add(HttpHeader.CONTENT_LENGTH, contentLength);
         HttpField ct = content.getContentType();
         if (ct != null)
-            response.getHeaders().put(ct);
+            headers.add(ct);
 
-        HttpField ce = content.getContentEncoding();
-        if (ce != null)
-            response.getHeaders().put(ce);
+        putHeaders(response);
+    }
 
-        if (_etags)
-        {
-            HttpField et = content.getETag();
-            if (et != null)
-                response.getHeaders().put(et);
-        }
+    protected void putNotModifiedHeaders(Response response, HttpContent content)
+    {
+        HttpFields.Mutable headers = response.getHeaders();
+        // send only lastModified, as it is too difficult to determine the etag with compression
+        HttpField lm = content.getLastModified();
+        if (lm != null)
+            headers.put(lm);
 
-        if (_acceptRanges && !response.getHeaders().contains(HttpHeader.ACCEPT_RANGES))
-            response.getHeaders().put(new PreEncodedHttpField(HttpHeader.ACCEPT_RANGES, "bytes"));
-        if (_cacheControl != null && !response.getHeaders().contains(HttpHeader.CACHE_CONTROL))
-            response.getHeaders().put(_cacheControl);
+        putHeaders(response);
+    }
+
+    protected void putHeaders(Response response)
```

Comment:
Make it `private`.

Reply:
protected so the service can be extended